0.21

balloob opened the pull request:
It's time for release 0.21 and it contains massive core improvements: replacement of our home grown HTTP stack with a standardized WSGI stack. This will improve performance, speed, security and make future development of advanced HTTP features a breeze.

This work was driven by the amazing Josh Wright. His knowledge, high standards and drive for security has helped improve Home Assistant a lot ever since he started helping out. Hip hip hurray for Josh!

Alright, time for the changes:
- New HTTP stack based on WSGI (@JshWright, @balloob)
- Frontend: lots of performance improvements (@balloob)
- Initial support for EnOcean lights, sensors and switches added (@rubund)
- Light: Osram Lightify is now supported (@olimpiurob)
- Light: Insteon Hub now supports brightness (@wkonkel)
- Add support for adding HA as Windows 10 tile (@fabaff)
- Condition: time condition can now cross midnight (@jaharkes)
- Template based sensors should now throw less warnings (@Bart247)
- LIRC component allows receiving IR commands (@partofthething)
- Feedreader will now persist data to disk to prevent duplicate commands (@shaftoe)
- Sun: azimuth attribute added (@fabaff)
- Flux like switch platform to change light intensity in the evening (@nkgilley)
- We no longer crash if you live in a part of the world where the sun never sets (@balloob)
- Rollershutter: RFXTRX now supported (@jacobtomlinson)
- Template switches can now execute scripts (@kellerza)
- Z-Wave: automatically heal the network at midnight (@infamy)
- Sensor: DTE Energy Bridge now supported (@kylehendricks)
- Media Player: Kodi now supports different turn off commands (@armills)
### Breaking Changes
- Our work in the WSGI stack is not fully done yet. We still have a minor issues where retrieving the error log in the about screen can raise an encoding error

Scheduled release date is Tuesday June 7

JshWright committed:
WIP: Add WSGI stack

This is a fair chunk of the way towards adding a WSGI compatible stack
for Home Assistant. The majot missing piece is auth/sessions. I was
undecided on implementing the current auth mechanism, or adding a new
mechanism (likely based on Werkzeug's signed cookies).

Plenty of TODOs...

JshWright committed:
Add some missing view registrations

balloob committed:
Fix import issues

balloob committed:
Tons of fixes - WIP

JshWright committed:
Improve view registration comments

Clarify that HomeAssistantWSGI.register_view() can handle either instantiated or uninstantiated view classes.

balloob committed:
More WIP

JshWright committed:
Friendlier exceptions for misconfigured views

If a view is missing a url or name attribute, this will result
in a more actionable exception being raised.

balloob committed:
Fix camera

balloob committed:
Add url validators

balloob committed:
Fix auth frontend

balloob committed:
Generate gzip for frontend/mdi

balloob committed:
Cache files in static folder for a year

balloob committed:
Access camera images using access token

balloob committed:
Get EventStream working

balloob committed:
Update requirements with new static update

balloob committed:
Cleanup

balloob committed:
Make event stream tests work on Travis ?

balloob committed:
Clean up EventStream

balloob committed:
Comment out eventstream tests

balloob committed:
Add gzip for static resources

balloob committed:
Gzip all the things

balloob committed:
Version bump to 0.21.0.dev0

balloob committed:
Remove gzip API

balloob committed:
Remove unused import

balloob committed:
Merge pull request #2063 from home-assistant/feature/wsgi

Feature/wsgi

balloob committed:
update frontend gz

balloob committed:
Another attempt to fix SSL in Docker

balloob committed:
EventStream to sent ping on start to notify browser

balloob committed:
Fix Dockerfile

balloob committed:
Use only 1 event listener for event stream.

jaharkes committed:
Ignore assertions from python threading when looking for leaked threads. (#2130)

While looking for leaked resources (threads) after shutdown and before restart
we in some cases get an assertion in the python threading module where we find
a thread marked as running at the python level but it has no associated thread
at the C level.

partofthething committed:
Added LIRC component for responding to IR remote commands

partofthething committed:
added LIRC component to .coveragerc

partofthething committed:
Removed LIRC dependency from requirements due to "complex" compliation

User will have to install lirc and python-lirc manually.

balloob committed:
Upgrade eventlet to 0.19

partofthething committed:
LIRC: Responded to some code review requests but not the big one

robbiet480 committed:
If no departure time is set, use now as the default. If departure time is set but does not have a :, assume its a preformed Unix timestamp and send along as raw input. Assume same for arrival_time.

robbiet480 committed:
If we have duration_in_traffic use that as the state, otherwise use duration

JshWright committed:
Fix TLS with eventlet (#2151)

* Fix TLS with eventlet

This fixes a simple error on my part when implementing the WSGI stuff.

eventlet.wrap_ssl() returns a wrapped socket, it does not modify the
object passed to it. We need to grab the returned value and use that.

* Fix style issue

jaharkes committed:
Don't even bother trying to kill stray child processes.

When we change our process group id we don't get keyboard interrupt
signals passed if our parent is a bash script.

partofthething committed:
Changed LIRC component so that it just fires events on the bus.

partofthething committed:
Moved lirc out of sensor package.

partofthething committed:
Updated requirements for LIRC

balloob committed:
manifest.json: remove trailing commas

wokar committed:
lg_netcast: fix exception on missing access_token (#2150)

* specified default value for acccess_token to prevent exception on init

balloob committed:
WSGI: Hide password in logs (#2164)

* WSGI: Hide password in logs

* Add auth + pw in logs tests

sxb1n9 committed:
Update PyISY to 1.0.6 (#2133)

* Update PyISY to 1.0.6 

1.0.6 is the newest version of PyISY

* PyISY to 1.0.6

fabaff committed:
Upgrade astral to 1.1 (#2131)

Commit:
Round temp and percentage for octoprint sensors (#2128)

partofthething committed:
Merge remote-tracking branch 'origin/dev' into lirc

partofthething committed:
Cleanups to LIRC module

Danielhiversen committed:
Fix bug in google travel time. Default option dictionary must contain mode. (#2134)

Commit:
Fixed issue with edimax SP-1101 switches (#2105)

* Fixed issue with edimax SP-1101 switches

* Added missing ValueError exception

partofthething committed:
Locked in required version of python-lirc for LIRC component

balloob committed:
Handle invalid dev ids for dev tracker + owntracks (#2174)

balloob committed:
Update frontend gzip

balloob committed:
Update frontend

balloob committed:
Update frontend

balloob committed:
Bugfixes for urls with dates

balloob committed:
Fix lint errors

balloob committed:
Update frontend

kellerza committed:
service helper replaced with script helper

balloob committed:
Update frontend

balloob committed:
Update frontend

Commit:
Add Vagrant setup (#2171)

balloob committed:
Update frontend

fabaff committed:
Vendorize vincenty requirement (#2176)

Commit:
Fixed roku exception when device is powered off or looses connection (#2173)

rubund committed:
New option for the netatmo platform: station (#2178)

This is necessary if multiple weather stations are associated with
one Netatmo account.

rubund committed:
Initial support for EnOcean (#2177)

* Initial support for EnOcean

Tested to work with:
 - Eltako FUD61 dimmer
 - Eltako FT55 battery-less switch
 - Permundo PSC234 (switch and power monitor)

* Rerun gen_requirements_all.py

olimpiurob committed:
Added Osram lightify platform (#2170)

* Added Osram Lightify light component

* Added color temperature and fade transition support to Osram Lightify

* Added Osram Lightify light component

* Added color temperature and fade transition support to Osram Lightify

* Updated docstring

* Added osramlightify to ignore list on coveragerc and updated docstrings

* Fixed linting issues

wkonkel committed:
Insteon support for brightness (#2169)

* Insteon support for brightness

* Farcy fix for unused constants.

* Remove unused constant and fix whitespace.

* Prevent toggle switches from jumping between states.

* 255 not 256

jaharkes committed:
Allow time condition windows to cross midnight. (#2158)

* Allow time condition windows to cross midnight.

* Address comments.

Fold _in_time_window back into the time() condition test.
Use specific time values to test the time window.

Commit:
Feedreader: add file data storage (#2147)

Right now we ignore already parsed entries and store the information
at runtime, but it will not survive a restart. This patch adds storage
functionality storing pickled file into default config folder when
feed has `published_parsed` support.

Bart274 committed:
List entity_ids in config and only react to them (#2144)

* List entity_ids in config and only react to them

This allows us to define a list of entity_ids in the config to make the
template sensor, binary sensor and switch only react to state changes of
these entities instead of listening to all state changes.

* Forgot to import the track_state_change function

* Changed test for added entity_ids to config

* Use default MATCH_ALL and remove event_listener

balloob committed:
Merge branch 'pr/2139' into dev

Conflicts:
	.coveragerc

fabaff committed:
Add azimuth (#1951)

* Add azimuth

* Place elevation and azimuth together in update part

balloob committed:
Fix linting issues

balloob committed:
Fix broken tests + linting

balloob committed:
Merge remote-tracking branch 'origin/master' into dev

Conflicts:
	homeassistant/const.py

rubund commented:
Cool.
I'm not so sure if we should brag so much about the EnOcean support yet. It works very well indeed with the three devices I have (Eltako FUD61,FT55, Permundo PSC234), but there are things that are hard-coded for them, and therefore one would most likely have to edit the code to get other devices working. It also lacks teaching of new sensors.

olimpiurob commented:
A somewhat similar situation for Osram Lightify.  There are RGBW bulbs,  white bulbs and LED strips.  I only have the RGBW bulb to play,  with so support may be lacking for white and LED strips.

balloob committed:
Improve index.html template

themactep committed:
Improve Internet Time calculation. (#2185)

partofthething commented:
> LIRC component allows sending and receiving IR commands (@partofthething)
Really it just allows receiving IR commands; sending is still accomplished via `shell_command`, which was possible before this change.

balloob commented:
I updated EnOcean to say initial support. As long as the documentation is clear about the currently supported devices, I think we're good.

This is pretty much how the Home Assistant development flows works -> someone adds initial support, someone else will test it and extend it for their device. If we don't announce it, that second person will not try it .

JshWright committed:
Re-add config validation for the http component (#2186)

This commit adds back the config validation for the http component. It
was removed during the WSGI shuffle. This is just a direct copy of what
@robbiet480 added in ab294d12f7 (with some testing to verify it still
works).

fabaff committed:
Remove print (already covered by logger) (#2184)

balloob committed:
Optimize foreacast.io API calls

fabaff committed:
Add windows 10 tile (#2166)

balloob committed:
Merge pull request #2191 from home-assistant/forecast-api-calls

Optimize foreacast.io API calls

robbiet480 committed:
Report whether SSL is required in discoverables, like /api/discovery_info and ZeroConf (#2192)

robbiet480 committed:
Revert "Report whether SSL is required in discovery"

balloob committed:
Merge pull request #2196 from home-assistant/revert-2192-expose-required-ssl-in-discoveries

Revert "Report whether SSL is required in discovery"

balloob committed:
Fix Norway (#2197)

nkgilley committed:
flux platform as a switch (#2097)

* flux platform as a switch

* use track_time_change. broken :(

* use track_utc_time_change instead of track_time_change

* add some basic tests

* use brightness from RGB_to_xy

* config_schema validation

* back to platform schema. what was i doing?

* more broken tests :(

* 644

* fix some time bugs

* add working tests. config validation still not right

* bug fixes and more test cases.

balloob committed:
Update frontend

infamy committed:
Update PULL_REQUEST_TEMPLATE.md (#2198)

* Update PULL_REQUEST_TEMPLATE.md

* Update PULL_REQUEST_TEMPLATE.md

jacobtomlinson committed:
Added rfxtrx rollershutter (#2030)

* Added rfxtrx rollershutter

* Updated mock command with real one

* Corrected test string

pavoni committed:
Bump pywemo version to fix ssdp discovery encoding issue.

pavoni committed:
Merge pull request #2201 from home-assistant/fix_pywemo_ssdp_decode_utf8

Bump pywemo version to fix ssdp discovery encoding issue.

pavoni committed:
Bump loopenergy version. Increased interval before deciding connection is dead and reconnecting.

pavoni committed:
Merge pull request #2203 from home-assistant/bump_loopenergy

Bump pyloopenergy version.

samriley committed:
Support for RFY protocol (#2199)

fabaff committed:
Upgrade python-telegram-bot to 4.2.0 (#2204)

kellerza committed:
Update Qwikswitch library version (#2214)

robbiet480 committed:
Add CORS to WSGI (#2209)

* Add CORS support to WSGI

* Remove X-HA-Access as a CORS header, because as @JshWright so elegantly put it: "CORS controls access to response headers, not request headers"

fabaff committed:
Upgrade blockchain to 1.3.3 (#2220)

kk7ds committed:
Add support for UniFi Video >= 3.2.0

Unfortunately, Ubiquiti changed their (supposedly versioned) API in
3.2.0 which causes us to have to refer to cameras by id instead of
UUID. The firmware for 3.2.x also changed the on-camera login procedures
and snapshot functionality significantly.

This bumps the requirement for uvcclient to 0.9.0, which supports the
newer API and makes the tweaks necessary to interact properly.

infamy committed:
Add a robots.txt (#2207)

fabaff committed:
Upgrade pysnmp to 4.3.2 (#2236)

* Upgrade pysnmp to 4.3.2

* Fix pylint issue

fabaff committed:
Upgrade slacker to 0.9.16 (#2235)

fabaff committed:
Upgrade fuzzywuzzy to 0.10.0 (#2234)

fabaff committed:
Upgrade schiene to 0.17 (#2231)

fabaff committed:
Add schema (#2226)

kellerza committed:
Updated template switch to cache Script objects

kellerza committed:
service helper replaced with script helper (#2242)

jabesq committed:
The metric unit of pressure is mbar not mBar (#2248)

This is useful to vaoid having several graph for the same type of data
According to wikipedia:
Units derived from the bar include the megabar (symbol: Mbar),
kilobar (symbol: kbar), decibar (symbol: dbar), centibar (symbol: cbar),
and millibar (symbol: mbar or mb).

JshWright commented:
CORS got merged in #2209

kk7ds committed:
Merge pull request #2241 from kk7ds/unifi-3.2

Add support for UniFi Video >= 3.2.0

Commit:
Improve vagrant provisioner resiliency (#2252)

This should make it easier to fix race conditions that might arise if
box is destroyed but setup_done placeholder file is not removed
properly

kylehendricks committed:
Add sensor for DTE Energy Bridge (#2247)

Currently only measures instantaneous energy usage in kW

turbokongen committed:
Hvac fix (#2221)

* Zwave hvac fix

* Zwave hvac fix and move max min temp to base

* Tests

infamy committed:
zwave auto heal at midnight (#2213)

* zwave auto heal at midnight

* fix debug to info, running heal, any heal will send a logger event

kellerza committed:
Default parameter for .run()

balloob committed:
Merge branch 'pr/2251' into dev

Conflicts:
	homeassistant/components/switch/template.py

emlove committed:
Add turn_off_action to kodi media player (#2224)

A new configuration option `turn_off_action` is added to kodi. It may be
one of: none, quit, hibernate, suspend, reboot, or poweroff. The
appropriate command is sent to kodi when the turn_off action is
requested. Default value is none.

Kodi will only report turn_off supported if it is configured to
something other than none.

Danielhiversen committed:
Fixed bug in google time travel (#2202)

Fixed bug in google time travel  when arrival time is given

balloob committed:
Version bump to 0.21

balloob committed:
Merge remote-tracking branch 'origin/master' into dev

Conflicts:
	homeassistant/const.py
